forgerock-openbanking-uk-aspsp-rs/forgerock-openbanking-uk-aspsp-rs-gateway/forgerock-openbanking-uk-aspsp-rs-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/rs/api/payment/verifier/DetachedJwsVerifier.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright 2019 ForgeRock AS.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.forgerock.openbanking.aspsp.rs.api.payment.verifier;
+
+import com.forgerock.openbanking.aspsp.rs.filter.MultiReadHttpServletRequest;
+import com.forgerock.openbanking.common.model.version.OBVersion;
+import com.forgerock.openbanking.common.services.store.tpp.TppStoreService;
+import com.forgerock.openbanking.exceptions.OBErrorException;
+import com.forgerock.openbanking.jwt.exceptions.InvalidTokenException;
+import com.forgerock.openbanking.jwt.services.CryptoApiClient;
+import com.forgerock.openbanking.model.Tpp;
+import com.forgerock.openbanking.model.error.OBRIErrorType;
+import com.nimbusds.jwt.JWTParser;
+import com.nimbusds.jwt.SignedJWT;
+import lombok.extern.slf4j.Slf4j;
+import net.logstash.logback.encoder.org.apache.commons.lang.StringUtils;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.security.Principal;
+import java.text.ParseException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.forgerock.openbanking.common.model.version.OBVersion.v3_1_4;
+import static com.forgerock.openbanking.model.error.OBRIErrorType.SERVER_ERROR;
+
+@Component
+@Slf4j
+public class DetachedJwsVerifier {
+
+    private static final Pattern JWS_DETACHED_SIGNATURE_PATTERN = Pattern.compile("(.*\\.)(\\..*)");
+
+    private final TppStoreService tppStoreService;
+    private final CryptoApiClient cryptoApiClient;
+
+    public DetachedJwsVerifier(TppStoreService tppStoreService, CryptoApiClient cryptoApiClient) {
+        this.tppStoreService = tppStoreService;
+        this.cryptoApiClient = cryptoApiClient;
+    }
+
+    public void verifyDetachedJws(String detachedJws, OBVersion obVersion, HttpServletRequest request, Principal principal) throws OBErrorException {
+        if (StringUtils.isEmpty(detachedJws)) {
```

Comment:
This is an important change and could have unexpected consequences. However, if verification of the detached JWS is enabled, surely it shouldn't be possible to send a null/empty one?!